pandas/core/sorting.py

```diff
@@ -285,6 +319,36 @@ def nargsort(
     return indexer
 
 
+def ensure_key_mapped(values, key: Optional[Callable]):
+    """
+    Applies a callable key function to the values function and checks
+    that the resulting value has the same shape. Can be called on Index
+    subclasses, Series, DataFrames, or ndarrays.
+
+    Parameters
+    ----------
+    values : Series, DataFrame, Index subclass, or ndarray
+    key : Optional[Callable], key to be called on the values array
+    """
+    if not key:
+        return values.copy()
+
+    _class = type(values)
+    result = key(values.copy())
+    if len(result) != len(values):
+        raise ValueError(
+            "User-provided `key` function much not change the shape of the array."
+        )
+
+    if not isinstance(result, _class):  # recover from type error
+        try:
+            result = _class(result)
+        except TypeError:
+            raise TypeError("User-provided `key` function returned an invalid type.")
+
+    return result
+
```

Comment:
I would move MultiIndex.apply_key here and make it a function that takes a MI and simply call this from ensure_key_mapper.

Reply:
My reasoning for keeping this outside of `ensure_key_mapped` is simply that we want `ensure_key_mapped` to be a generic function that takes any object and tries to apply a key function to it. My feeling for the API is that calling functions should be responsible for handling any kind of preprocessing they want to do, instead of cluttering it with `isinstance` checks. I also though the `MultiIndex.apply_key` was a useful function, sort of like map that allows you to transform a `MultiIndex`. If you think this is best, I'm happy to do it though.

Reply:
I don think its a generally useful function and is very specific to this case. we want to move all unecessary logic out of the sorting functions.

Reply:
Ok I moved it to `sorting.py` and added a ton of documentation. And dealt with some crazy linting issues.